src/libOpenImageIO/imagebufalgo_xform.cpp

```diff
@@ -1144,4 +1155,255 @@ ImageBufAlgo::rotate(const ImageBuf& src, float angle, string_view filtername,
+                     chan_t);
+        return false;
+    }
+    // N.B. We currently require floating-point ST channels
+    if (!stSpec.format.is_floating_point()) {
+        dst.error("ImageBufAlgo::st_warp : ST buffer must be holding floating-"
+                  "point data");
+        return false;
+    }
+
+    if (dst.initialized()) {
+        // XXX: Currently requiring the pixel scales of ST and a preallocated
+        // output buffer to match.
+        if (dst.spec().full_width != stSpec.full_width
+            || dst.spec().full_height != stSpec.full_height) {
+            dst.error("ImageBufAlgo::st_warp : Output and ST buffers must have "
+                      "the same full width and height");
+            return false;
+        }
+    }
+
+    // Prep the dest spec using the channels from `src`, and the intersection of
+    // `roi` and the ROI from the ST buffer (since the ST warp is only defined
+    // for pixels in the ST buffer). We grab a copy of the input ROI before
+    // `IBAprep`, since we want to intersect it ourselves.
+    ROI inputROI(roi);
+    bool res
+        = ImageBufAlgo::IBAprep(roi, &dst, &src,
+                                ImageBufAlgo::IBAprep_NO_SUPPORT_VOLUME
+                                    | ImageBufAlgo::IBAprep_NO_COPY_ROI_FULL);
+    if (res) {
+        roi         = roi_intersection(inputROI, stSpec.roi());
+        roi.chbegin = std::max(roi.chbegin, 0);
+        roi.chend   = std::min(roi.chend, src.spec().nchannels);
+
+        ImageSpec destSpec(dst.spec());
+        destSpec.set_roi(roi);
+        destSpec.set_roi_full(stSpec.roi_full());
+        // XXX: It would be nice to be able to tell `IBAprep` to skip the buffer
+        // initialization. Worth adding a new prep flag?
+        dst.reset(destSpec);
+    }
+    return res;
+}
+
+
+
+bool
+ImageBufAlgo::st_warp(ImageBuf& dst, const ImageBuf& src, const ImageBuf& stbuf,
+                      Filter2D* filter, int chan_s, int chan_t, bool flip_s,
+                      bool flip_t, ROI roi, int nthreads)
+{
+    pvt::LoggedTimer logtime("IBA::st_warp");
+
+    if (!check_st_warp_args(dst, src, stbuf, chan_s, chan_t, roi)) {
+        return false;
+    }
+
+    // Set up a shared pointer with custom deleter to make sure any
+    // filter we allocate here is properly destroyed.
+    std::shared_ptr<Filter2D> filterptr((Filter2D*)nullptr, Filter2D::destroy);
+    if (!filter) {
+        // If a null filter was provided, fall back to a reasonable default.
+        filterptr.reset(Filter2D::create("lanczos3", 6.0f, 6.0f));
+        filter = filterptr.get();
+    }
+
+    bool ok;
+    OIIO_DISPATCH_COMMON_TYPES3(ok, "st_warp", st_warp_, dst.spec().format,
+                                src.spec().format, stbuf.spec().format, dst,
+                                src, stbuf, chan_s, chan_t, flip_s, flip_t,
+                                filter, roi, nthreads);
```

Comment:
`OIIO_DISPATCH_COMMON_TYPES3` will handle all combinations of the types of the three images. But if I understand correctly, you already disallow stbuf being anything other than float. So that means that `OIIO_DISPATCH_COMMON_TYPES2` should work and will make many fewer template expansions. (Or else, stick with 3, and then you can remove the restriction on the type of stbuf, and I think it should all work.)

Reply:
Thanks, good catch. I originally thought it would be important to maintain the exact pixel datatype in the ST buffer when computing the lookup coordinates, but since I decided to punt and just use float across the board, I'll definitely change this.

Reply:
OK, after taking another look at this, I'm not sure I can trivially switch to `OIIO_DISPATCH_COMMON_TYPES2` here.

> But if I understand correctly, you already disallow stbuf being anything other than float.

I only check that `stbuf` is a floating-point datatype, so `half`, `float`, and `double` are all acceptable. Thus, I can't just convert the `stbuf` iterator to `ImageBuf::ConstIterator<float>`, since that will misrepresent the buffer's source data and lead to garbage results (e.g. when using half ST data).

I'm not sure there's a way out of this without potentially adding another special-purpose dispatch template/macro, but let me know what you think.

Reply:
Ah, I see now. You're right, I misread the is_float_point as a hard test for float.

If you use COMMON_TYPES3, then it will automatically handle any time for stbuf, it doesn't really have to be float-like.

Reply:
...automatically handle any TYPE for stbuf...

Reply:
I just realized that I think I misunderstood how the auto-conversion done by `ImageBuf::Iterator` works between fixed- and floating-point types...

The extra `is_floating_point` check I've been doing was so I could avoid having to normalize fixed-point values from the ST buffer, but I see now that the `conver_type` helpers look to be scaling fixed-point values to/from the normalized floating-point range (at least for unsigned fixed-point types). Thus, I think that extra restriction can go, and I can just let accept any datatype for `stbuf`.

Reply:
Yes, exactly, the normalization of integer types to [0.0-1.0] floating point values automatically happens when you dereference the iterator, that's why it's templated on the underlying data type! You can, in effect, write your "kernel" of an IBA function as if everything were floating point, and the iterators handle the conversion in and out (at the expense of the big template expansion handled by the `OIIO_DISPATCH_*` macros).